packages/votemarket/src/oracle/PendleOracle.sol

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.19;
+
+///  Project Interfaces
+import "src/interfaces/IPendleOracle.sol";
+
+/// @notice Oracle contract to store voting values from the Gauge Controller.
+contract PendleOracle {
+    ////////////////////////////////////////////////////////////////
+    /// --- DATA STRUCTURE DEFINITIONS
+    ///////////////////////////////////////////////////////////////
+
+    /// @notice Curve Gauge Controller Structs appended with last update timestamp.
+    struct Point {
+        uint256 bias;
+        uint256 lastUpdate;
+    }
+
+    struct VotedSlope {
+        uint256 slope;
+        uint256 end;
+        uint256 lastUpdate;
+    }
+
+    ////////////////////////////////////////////////////////////////
+    /// --- STORAGE VARIABLES
+    ///////////////////////////////////////////////////////////////
+
+    /// @notice Governance address.
+    address public governance;
+
+    /// @notice Future governance address.
+    address public futureGovernance;
+
+    /// @notice Mapping of addresses authorized to insert data into the contract.
+    mapping(address => bool) public authorizedDataProviders;
+
+    /// @notice  Mapping of addresses authorized to insert block numbers into the contract.
+    mapping(address => bool) public authorizedBlockNumberProviders;
```

Comment:
For DX and gas efficiently in some scenarios, instead of using a mapping per permission, you can leverage the `enum` type to create a permissions mapping.

```solidity
    // Enum representing shipping status
    enum Permissions {
        None,
        DATA,
        BLOCK_NUMBER,
        ALL,
    }

    mapping(address => Permissions) public permissions;
 ```
 
Feel free to take into account this comment or not. Keep just in mind that Enum are just sugar-syntax in Solidity.